### ISSUE TITLE
High level client API

### ISSUE THREAD

**Issue by phillip-kruger** [eclipse/microprofile-graphql#185]
Linked to #96 - We need a High Level Client API

**Comment by t1** [comment]
I think there are two options for the high level client api, and we may want to provide both:

a) schema-first: generate the types and query/mutation methods with a build tool (e.g. a maven plugin or annotation processor). This is what the other tools do and what probably is most natural for GraphQL.

b) code-first: map types and query/mutation methods written in Java to appropriate GraphQL requests and responses. Consumer Driven Contracts are an interesting way of defining APIs; maybe it's also suitable for GraphQL that's worth investigating further. I did some experiments and they look good so far.

What do you think?

**Comment by t1** [comment]
I continued with my experiment of how a consumer contract high level api could look like. It doesn't support any annotations, yet, but it might be worth taking a look: https://github.com/t1/graphql-java-experiment/blob/master/src/test/java/it/SuperHeroesIT.java

**Comment by phillip-kruger** [comment]
I like it. It does mean we might have to move some annotations (like `@Query` and `@Mutation`) to a common module, so that we don't have the client depend on the server.

**Comment by t1** [comment]
I just added a very basic support for the `@Query` annotation, so you can have multiple query methods returning classes with different fields.

**Comment by andymc12** [comment]
I like it too.  If we want to be consistent with the MP Rest Client (not a requirement), then we should put the class in the `build` method, and rename "builder" to "newBuilder" like this:

```
    private final SuperHeroesApi api = GraphQlClient.newBuilder()
        .endpoint("http://localhost:8080/graphql-java-experiment/graphql")
        .build(SuperHeroesApi.class);
```

We'll probably need to pull over other annotations into a common module - I could see `@Name` being useful with this type of API - for arguments as well as queries, etc.

I like the simplicity of this and the consistency with the MP Rest Client.  Thanks!

**Comment by t1** [comment]
I just changed the syntax as suggested.

Should I extract it into a separate module. I also don't feel well to still work without dedicated tests.

**Comment by phillip-kruger** [comment]
Hi @t1 - It's awesome that you started thinking about the high-level client. At the moment it's targeted for a future release, so I would suggest that we do the POC/your work in a separate project. It will be good to "test" the implementation before we come back to put it in the spec. If you agree we can maybe talk about getting this into smallrye ? Let me know

**Comment by t1** [comment]
I moved it to https://github.com/t1/graphql-client for now... smallrye sounds like a good path to consider.

**Comment by t1** [comment]
Done with the CDI extension, so you can `@Inject` an interface annotated as `@GraphQlClientApi`. See [here](https://github.com/t1/graphql-java-experiment/blob/d1f4bf890ddfc52605b2e98246e994b4c4d32bae/src/main/java/org/superheroes/team/TeamsBoundary.java#L25) for an example.

I would be glad to hear your feedback.

**Comment by andymc12** [comment]
Hi @t1 - I like it - it looks very similar to the MP Rest Client.  Would you then use MP Config to specify the URL endpoint for the remote GraphQL service?  Or maybe have a `url` field in the `@GraphQlClientApi` annotation that could be overridden by MP Config?

**Comment by t1** [comment]
I must admit that it's a shameless plug from MP Rest Client ;-) To express it more positively: I tried to align it to MPRC as much as possible. So, yes, you can configure the endpoint with a MP Config `<prefix>/mp-graphql/url` or in the `@GraphQlClientApi` annotation, where the default prefix is the interface name that can be set with the `configKey` field in that annotation.

But it uses only one `@GraphQlClientApi` annotation on the interface; no need to `@RegisterQlClientApi` on the interface plus qualifier `@GraphQlClientApi` on the injection point. I didn't see the point in that.

**Comment by andymc12** [comment]
> I must admit that it's a shameless rip off from MP Rest Client ;-) To express it more positively: I tried to align it to MPRC as much as possible. So, yes, you can configure the endpoint with a MP Config <prefix>/mp-graphql/url or in the @GraphQlClientApi annotation, where the default prefix is the interface name that can be set with the configKey field in that annotation.

I think consistency in the MP specs is definitely a good thing.  Keep ripping off from other specs!!  ;-)

> But it uses only one @GraphQlClientApi annotation on the interface; no need to @RegisterQlClientApi on the interface plus qualifier @GraphQlClientApi on the injection point. I didn't see the point in that.

MP Rest Client used `@RegisterRestClient` (on the interface) and `@RestClient` (on the injection target) as a qualifier.  I don't know if the limitation still exists or not, but at one time, CDI didn't handle the injection of Java proxies too well without a qualifier.  If that is no longer required, then let's keep it simple, but if things don't work without the qualifier, we may need to put the qualifier on the injection target.

**Comment by t1** [comment]
@andymc12: It works on WildFly using Weld. I didn't test it on other containers, but IIRC the simpler approach should be spec compliant.

**Comment by andymc12** [comment]
@t1, I showed off your source code - the example and some of the implementation code in our weekly meeting, and it seemed very popular!  Would you be willing to demo it at next week's meeting?  We meet on Fridays at 10:30am US Eastern Time (all of the MP project meetings are available on [this calendar](https://calendar.google.com/calendar?cid=Z2JuYmMzNzNnYTQwbjB0dmJsODhua2MzcjRAZ3JvdXAuY2FsZW5kYXIuZ29vZ2xlLmNvbQ)). Thanks!

**Comment by t1** [comment]
Yes, sure! I feel honored to get invited :-)

**Comment by andymc12** [comment]
I'm glad to hear it!  Also, we have a "running chat" on gitter - issues and the weekly meetings are the more official ways fo communication, but sometimes it's nice to ask quick questions or have short conversations with everybody.  You can join here:  https://gitter.im/eclipse/microprofile-graphql

**Comment by pilhuhn** [comment]
Haven't looked at the example test yet. I think (allowing for) schema first is a must. In my case I e.g. have a server in Python, that someone provides me and I need to talk to it from my Java client. Manually modelling the Java classes in the hope that they match the schema makes little sense to me.

I think the litmus test is :   create some domain model, expose it via GraphQL server + generated schema. Use client tools to create client classes from the schema. Those client classes should match the server's domain model.

**Comment by phillip-kruger** [comment]
Hi @pilhuhn 
We already have some code to show how you can generate your client. We decided that generation should not be part of the spec. It might be part of an implementation to generate code that looks like the defined code in the spec.

Mostly, when using graphql, you do NOT want the server domain model (else you might as well use REST ?). You want your own subset, that fetch only what you wants. 

Here is a POC of a client generator that we are playing with in SmallRye: https://github.com/smallrye/smallrye-graphql/pull/215

**Comment by pilhuhn** [comment]
@phillip-kruger Yes, I get that you (most of the time) don't want a 1:1 exact model. And I don't have enough knowledge yet to word it well enough. What I wanted to come to is that things only make sense if a Client domain-model can be bootstrapped from a schema.

The spec may not need to tell how it can be done, but should ensure that an implementation does it in a consistent way. E.g. by also applying annotations (where they make sense) that are the same as on the server side (domain model) to reduce the learning effort.

Wrt REST vs GraphQL: I believe that if someone starts a new project with GraphQL, they may not want to build a parallel infrastructure via REST.

**Comment by t1** [comment]
@pilhuhn : If you are in a DDD Conformist position, a perfect solution would be to have an IDE with full support for auto-completion and documentation derived from a schema; something like GraphiQL but directly producing and executing Java code. As long as we don't have that, we have to learn the Query language, use a tool like GraphiQL, and then maybe work with some code generator to produce the Java code. The PoC mentioned by @phillip-kruger supports this workflow; maybe this is what you need.

But this is very different from the workflow I see sometimes in Rest services, where the Service team also produces an API jar containing the DTOs and maybe even the MP Rest Client API interface, so a client can use it directly. This makes sense for Rest services, because the full data set will be produced by the service and sent to the client anyway. The client then only uses the fields it needs. But for GraphQL services every field should be an active decision from the client.
